features/slack/lib.php

```diff
@@ -0,0 +1,323 @@
+<?php
+
+include 'slack/lib/message_formatter.php';
+include 'slack/lib/date_time_util.php';
+
+class Slack {
+
+    /** constants mapped from Persistence class */
+    const OK = Persistence::OK;
+    const ERROR = Persistence::ERROR;
+
+    private static $user_list;
+
+    private $oAuthToken;
+    private $curlClient;
+
+    function __construct(CurlClient $curlClient) {
+        $this->curlClient = $curlClient;
+
+        $slack_config = Configuration::get_configuration("slack");
+        $this->oAuthToken = array_key_exists('oAuth_token', $slack_config) ? $slack_config['oAuth_token']: '';
+    }
+
+    /**
+     * get a list of the Slack Users
+     *
+     * @returns array of Slack_ user list
+     */
+    function get_user_list() {
+
+        // Load user list once for a request (process)
+        if(isset(self::$user_list)) {
+            return self::$user_list;
+        }
+
+        $apiResult = $this->curlClient->get('https://etsy.slack.com/api/users.list', array('token' => $this->oAuthToken));
+        $userListData = json_decode($apiResult);
+
+        $userList = [];
+        if($userListData->ok != true ) {
+            return $userList;
+        }
+
+        foreach ($userListData->members as $userInfo) {
+            $id = $userInfo->id;
+            $userList[$id] = $userInfo;
+        }
+        self::$user_list=$userList;
+        return self::$user_list;
+    }
+
+
+    /**
+     * get a list of the Slack channels that can be selected for a given postmortem
+     * if the 'morgue_get_slack_channels_list' exists, call it and return
+     * its results - otherwise, lookup the config file for ['Slack_']['channels']
+     *
+     * @returns array of Slack_ channels strings
+     */
+    function get_slack_channels_list() {
+        if (function_exists("morgue_get_slack_channels_list")) {
+            return morgue_get_slack_channels_list();
+        } else {
+            $queryParam = array(
+                'token'             => $this->oAuthToken,
+                'exclude_archived'  => 1,
+                'exclude_members'   => 1
+            );
+
+            $result = $this->curlClient->get('https://slack.com/api/channels.list', $queryParam);
+
+            // If the API is JSON, use json_decode.
+            $data = json_decode($result);
+            $channelList = [];
+            if($data->ok == true ) {
+                foreach ($data->channels as $channelInfo) {
+                    $channelId  = $channelInfo->id;
+                    $channelName= $channelInfo->name;
+                    $channelList[$channelId] = $channelName;
+                }
+            }
+            return $channelList;
+        }
+    }
+
+    /**
+     * @param $starttime
+     * @param $endtime
+     * @param $channel_id
+     * @return string
+     */
+    function get_channel_messages_for_datetime_range($starttime, $endtime, $channel_id) {
```

Comment:
might want to set a maximum number of requests to make, or messages to get. in case this receives an unexpectedly large time-range. Even a 24 hour period in one channel could be thousands of messages.

Reply:
Added maximum number of request limit.